### PR TITLE
Fix for defect #695: permission check for multibranch pipelines

### PR DESCRIPTION
**Pull request by mlynnyk**
This pull request addresses this issue: https://github.com/jenkinsci/gitlab-plugin/issues/695 . The discussion does provide a detailed description of the defect.

While it is not a critical security problem, the issue does, in fact, results in allowing a user to perform branch indexing action, which is normally restring by the "Job/Build" permission.

Having the plugin with the code change installed, and trying to trigger a webhook against a multibranch pipeline would return not 200 (which is current behavior in all cases) but the correct "Error 403 anonymous is missing the Job/Build permission" response. Likewise, having a Webhook configured with basic auth ( http://username:apitoken@my_jenkins.com:8080/project/pipeline_test/ ) when user "username" does not have sufficient permissions, would result in "Error 403 username is missing the Job/Build permission", which, again, should be the correct behavior.

### CODE REVIEW

**Comment by krisstern** [comment]
Thanks @mlynnyk for you code contribution! I will review your PR over the next few days and merge once it is approved and we are ready to do so.

File: src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java [hunk 1]
```diff
@@ -122,5 +132,25 @@ public void run() {
                 }
             }
         }
+
+        private void checkPermission(Permission permission) {
+            GitLabConnectionConfig gitlabConfig =
+                    (GitLabConnectionConfig) Jenkins.get().getDescriptor(GitLabConnectionConfig.class);
+            if (gitlabConfig != null) {
+                if (gitlabConfig.isUseAuthenticatedEndpoint()) {
+                    if (!project.getACL().hasPermission(authentication, permission)) {
```
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java:141]: 'hasPermission(org.acegisecurity.Authentication, hudson.security.Permission)' is deprecated, suggest using https://javadoc.jenkins-ci.org/hudson/security/AccessControlled.html#hasPermission2(org.springframework.security.core.Authentication,hudson.security.Permission) instead.

File: src/test/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildActionTest.java [hunk 2]
```diff
@@ -102,10 +103,13 @@ private String getJson(String name) throws IOException {
     public void scmSourceOnUpdateExecuted() {
         GitSCMSource source = new GitSCMSource("http://test");
         SCMSourceOwner item = mock(SCMSourceOwner.class);
+        ACL acl = mock(ACL.class);
         when(item.getSCMSources()).thenReturn(Collections.singletonList(source));
+        when(item.getACL()).thenReturn(acl);
         Assert.assertThrows(
                 HttpResponses.HttpResponseException.class,
                 () -> new PushBuildAction(item, getJson("PushEvent.json"), null).execute(response));
+        item.onSCMSourceUpdated(source);
```
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildActionTest.java:112]: 'onSCMSourceUpdated(jenkins.scm.api.SCMSource)' is deprecated, implementations of [SCMSourceOwner](https://javadoc.jenkins.io/plugin/scm-api/jenkins/scm/api/SCMSourceOwner.html) would prefer the [SCMEventListener](https://javadoc.jenkins.io/plugin/scm-api/jenkins/scm/api/SCMEventListener.html) extension point which allows for more fine-grained response to events, so prefer delivering event notification through [SCMHeadEvent.fireNow(SCMHeadEvent)](https://javadoc.jenkins.io/plugin/scm-api/jenkins/scm/api/SCMHeadEvent.html#fireNow(jenkins.scm.api.SCMHeadEvent)), [SCMSourceEvent.fireNow(SCMSourceEvent)](https://javadoc.jenkins.io/plugin/scm-api/jenkins/scm/api/SCMSourceEvent.html#fireNow(jenkins.scm.api.SCMSourceEvent)) or [SCMNavigatorEvent.fireNow(SCMNavigatorEvent)](https://javadoc.jenkins.io/plugin/scm-api/jenkins/scm/api/SCMNavigatorEvent.html#fireNow(jenkins.scm.api.SCMNavigatorEvent)) as appropriate.

File: src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java [hunk 3]
```diff
@@ -90,13 +93,19 @@ protected void performOnPost(GitLabPushTrigger trigger) {
...
         if (project instanceof SCMSourceOwner) {
-            ACL.impersonate(ACL.SYSTEM, new SCMSourceOwnerNotifier());
+            ACL.impersonate(ACL.SYSTEM, new SCMSourceOwnerNotifier(Jenkins.getAuthentication()));
             throw HttpResponses.ok();
         }
         throw HttpResponses.errorWithoutStack(409, "Push Hook is not supported for this project");
     }
 
     private class SCMSourceOwnerNotifier implements Runnable {
+        private final Authentication authentication;
```
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java:103]: 'org.acegisecurity.Authentication' is deprecated.

File: src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java [hunk 4]
```diff
@@ -90,13 +93,19 @@ protected void performOnPost(GitLabPushTrigger trigger) {
...
+
+        public SCMSourceOwnerNotifier(Authentication authentication) {
```
Comment (krisstern) [inline src/main/java/com/dabsquared/gitlabjenkins/webhook/build/PushBuildAction.java:105]: 'org.acegisecurity.Authentication' is deprecated.

**Review by krisstern — APPROVED** [review]
Aside from some deprecation issues, the PR looks okay.

**Comment by krisstern** [comment]
@mlynnyk Would you like to address the deprecations, or are you happy with the code as is?

**Comment by mlynnyk** [comment]
Hi @krisstern , thank you for reviewing this and providing your feedback. I would say that the matter of addressing these deprecation issues appears to be somewhat wider than this particular code change. For example, when replacing "org.acegisecurity.Authentication" with "org.springframework.security.core.Authentication" (I believe that's the correct one) it would make sense to update other classes using it and eliminate its usage completely.

When I added this PR, I merely wanted to fix the defect while keeping the code more or less consistent with the rest of the implementation. I don't really want to get involved with a wider refactoring, this appears to be a whole separate task. Also, I am currently running the plugin with this code change on a production environment and it works perfectly well, so it has been tested. Because of these reasons I would prefer to leave this code as it is now.

**Comment by krisstern** [comment]
Hi @mlynnyk Sure, let me check things interactively one more time before merging this. Again, thank you for your contribution!

**Comment by MattLud** [comment]
Hi there - as a follow up to this PR, does the existing Multibranch documentation still apply? 

It seems like tokens won't be used going forward (nor were they ever used?) and the official doc should request use of Basic Auth parameters. 

https://github.com/jenkinsci/gitlab-plugin?tab=readme-ov-file#pipeline-multibranch-jobs-1

**Pull request merged**